src/coredump/coredump.c

```diff
@@ -1328,6 +1329,257 @@ static int gather_pid_metadata_from_procfs(struct iovec_wrapper *iovw, Context *
+        cmsg->cmsg_len = CMSG_LEN(sizeof(struct ucred));
+        memcpy(CMSG_DATA(cmsg), ucred, sizeof(struct ucred));
+
+        cmsg = CMSG_NXTHDR(&mh, cmsg);
+        cmsg->cmsg_level = SOL_SOCKET;
+        cmsg->cmsg_type = SCM_RIGHTS;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(int));
+        memcpy(CMSG_DATA(cmsg), &fd, sizeof(int));
+
+        return RET_NERRNO(sendmsg(transport_fd, &mh, MSG_NOSIGNAL));
+}
+
+static int receive_ucred_and_one_fd(int transport_fd, struct ucred *ret_ucred, int *ret_fd) {
+        _cleanup_close_ int fd = -EBADF;
+        CMSG_BUFFER_TYPE(CMSG_SPACE(sizeof(struct ucred)) +
+                         CMSG_SPACE(sizeof(int))) control = {};
+        struct msghdr mh = {
+                .msg_control = &control,
+                .msg_controllen = sizeof(control),
+        };
+        struct cmsghdr *cmsg = NULL;
+        struct ucred *ucred = NULL;
+        ssize_t n;
+        int r;
+
+        assert(ret_fd);
+        assert(ret_ucred);
+
+        r = setsockopt_int(transport_fd, SOL_SOCKET, SO_PASSCRED, true);
+        if (r < 0)
+                return r;
+
+        n = recvmsg_safe(transport_fd, &mh, 0);
+        if (n < 0)
+                return n;
+
+        CMSG_FOREACH(cmsg, &mh)
+                if (cmsg->cmsg_level == SOL_SOCKET &&
+                    cmsg->cmsg_type == SCM_RIGHTS &&
+                    cmsg->cmsg_len == CMSG_LEN(sizeof(int))) {
+
+                        assert(fd < 0);
+                        fd = *CMSG_TYPED_DATA(cmsg, int);
+
+                } else if (cmsg->cmsg_level == SOL_SOCKET &&
+                           cmsg->cmsg_type == SCM_CREDENTIALS &&
+                           cmsg->cmsg_len == CMSG_LEN(sizeof(struct ucred))) {
+
+                        assert(!ucred);
+                        ucred = CMSG_TYPED_DATA(cmsg, struct ucred);
+                }
+
+        if (!ucred || fd < 0)
+                return -EIO;
+
+        *ret_fd = TAKE_FD(fd);
+        *ret_ucred = *ucred;
+
+        return 0;
+}
+
+static int can_forward_coredump(pid_t pid) {
+        _cleanup_free_ char *path = NULL, *unit = NULL;
+        pid_t ppid;
+        int r;
+
+        r = get_process_ppid(pid, &ppid);
+        if (r < 0)
+                return r;
+
+        r = cg_pid_get_path(SYSTEMD_CGROUP_CONTROLLER, ppid, &path);
+        if (r < 0)
+                return r;
+
+        r = cg_path_get_unit_path(path, &unit);
+        if (r == -ENOMEM)
+                return log_oom();
+        if (r == -ENXIO)
+                /* No valid units in this path. */
+                return false;
+        if (r < 0)
+                return r;
+
+        /* We require that this process belongs to a delegated cgroup
+         * (i.e. user.delegate=1), with user.coredump_receive=1 also. */
+        r = cg_is_delegated(SYSTEMD_CGROUP_CONTROLLER, unit);
+        if (r <= 0)
+                return r;
+
+        r = cg_get_xattr_bool(SYSTEMD_CGROUP_CONTROLLER, unit, "user.coredump_receive");
```

Comment:
i kinda like the cg_is_delegated() thing you added. Maybe we should add a similar wrapper for the coredump receive stuff, i.e. hide the xattr away inside a function.